substrate/client/rpc-servers/src/middleware/chain_head.rs

```diff
@@ -0,0 +1,237 @@
+					};
+				}
+			}
+		}
+
+		ResponseFuture::Forward { fut: self.service.call(req.clone()) }
+	}
+}
+
+/// Extract the subscription ID from the provided parameters.
+///
+/// We make the assumption that all `chainHead` methods are given the
+/// subscription ID as a first parameter.
+///
+/// This method handles positional and named `camelCase` parameters.
+fn get_subscription_id<'a>(params: Params<'a>) -> Option<String> {
+	// Support positional parameters.
+	if let Ok(follow_subscription) = params.sequence().next::<String>() {
+		return Some(follow_subscription)
+	}
+
+	// Support named parameters.
+	let Ok(value) = params.parse::<serde_json::Value>() else { return None };
+
+	let serde_json::Value::Object(map) = value else { return None };
+	if let Some(serde_json::Value::String(subscription_id)) = map.get("followSubscription") {
+		return Some(subscription_id.clone())
+	}
+
+	None
+}
+
+/// Extract the result of a jsonrpc object.
+///
+/// The function extracts the `result` field from the JSON-RPC response.
+///
+/// In this example, the result is `tfMQUZekzJLorGlR`.
+/// ```ignore
+/// "{"jsonrpc":"2.0","result":"tfMQUZekzJLorGlR","id":0}"
+/// ```
+fn get_method_result(response: &MethodResponse) -> Option<String> {
+	if response.is_error() {
+		return None
+	}
+
+	let result = response.as_result();
+	let Ok(value) = serde_json::from_str(result) else { return None };
+
+	let serde_json::Value::Object(map) = value else { return None };
+	let Some(serde_json::Value::String(res)) = map.get("result") else { return None };
+
+	Some(res.clone())
+}
+
+/// Response future for chainHead middleware.
+#[pin_project(project = ResponseFutureProj)]
+pub enum ResponseFuture<F> {
+	/// The response is propagated immediately without calling other layers.
+	///
+	/// This is used in case of an error.
+	Ready {
+		/// The response provided to the client directly.
+		///
+		/// This is `Option` to consume the value and return a `MethodResponse`
+		/// from the projected structure.
+		response: Option<MethodResponse>,
+	},
+
+	/// Forward the call to another layer.
+	Forward {
+		/// The future response value.
+		#[pin]
+		fut: F,
+	},
+
+	/// Forward the call to another layer and store the subscription ID of the response.
+	Register {
+		/// The future response value.
+		#[pin]
+		fut: F,
+		/// Connection data that captures the subscription ID.
+		connection_data: Arc<Mutex<ConnectionData>>,
+	},
+}
+
+impl<'a, F> std::fmt::Debug for ResponseFuture<F> {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		f.write_str("ResponseFuture")
+	}
+}
+
+impl<F: Future<Output = MethodResponse>> Future for ResponseFuture<F> {
+	type Output = F::Output;
+
+	fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+		let this = self.project();
+
+		match this {
+			ResponseFutureProj::Ready { response } =>
+				Poll::Ready(response.take().expect("Value is set; qed")),
+			ResponseFutureProj::Forward { fut } => fut.poll(cx),
```

Comment:
why are you not removing subscriptions when `chainHead_unstable_unfollow` are called?